ocaml/typing/ctype.ml

```diff
@@ -4576,71 +4581,119 @@ let is_moregeneral env inst_nongen pat_sch subj_sch =
   | () -> true
   | exception Moregen _ -> false
 
-(* Alternative approach: "rigidify" a type scheme,
-   and check validity after unification *)
-(* Simpler, no? *)
-
-let rec rigidify_rec vars ty =
-  if try_mark_node ty then
-    begin match get_desc ty with
-    | Tvar _ ->
-        if not (TypeSet.mem ty !vars) then vars := TypeSet.add ty !vars
-    | Tvariant row ->
-        let Row {more; name; closed} = row_repr row in
-        if is_Tvar more && not (has_fixed_explanation row) then begin
-          let more' = newty2 ~level:(get_level more) (get_desc more) in
-          let row' =
-            create_row ~fixed:(Some Rigid) ~fields:[] ~more:more'
-              ~name ~closed
-          in link_type more (newty2 ~level:(get_level ty) (Tvariant row'))
-        end;
-        iter_row (rigidify_rec vars) row;
-        (* only consider the row variable if the variant is not static *)
-        if not (static_row row) then
-          rigidify_rec vars (row_more row)
-    | _ ->
-        iter_type_expr (rigidify_rec vars) ty
-    end
-
-let rigidify ty =
-  let vars = ref TypeSet.empty in
-  rigidify_rec vars ty;
-  unmark_type ty;
-  TypeSet.elements !vars
-
 let all_distinct_vars env vars =
   let tys = ref TypeSet.empty in
   List.for_all
     (fun ty ->
       let ty = expand_head env ty in
-      if TypeSet.mem ty !tys then false else
-      (tys := TypeSet.add ty !tys; is_Tvar ty))
+      if TypeSet.mem ty !tys then false else begin
+       tys := TypeSet.add ty !tys;
+       is_Tvar ty
+     end)
     vars
 
-let matches ~expand_error_trace env ty ty' =
-  let snap = snapshot () in
-  let vars = rigidify ty in
-  cleanup_abbrev ();
-  match unify env ty ty' with
-  | () ->
-      if not (all_distinct_vars env vars) then begin
-        backtrack snap;
-        let diff =
-          if expand_error_trace
-          then expanded_diff env ~got:ty ~expected:ty'
-          else unexpanded_diff ~got:ty ~expected:ty'
-        in
-        raise (Matches_failure (env, unification_error ~trace:[diff]))
-      end;
-      backtrack snap
-  | exception Unify err ->
+type matches_result =
+  | Unification_failure of Errortrace.unification_error
+  | Layout_mismatch of { original_layout : layout; inferred_layout : layout
+                       ; ty : type_expr }
+  | All_good
+
+module Matches = struct
+  (* Alternative approach: "rigidify" a type scheme,
+     and check validity after unification *)
+  (* Simpler, no? *)
+
+  let rec rigidify_rec vars ty =
```

Comment:
I think moving this function and its friends into a module is probably a mistake for rebasing reasons.  The functions are mostly unchanged, but now every line has a diff because of whitespace.